### DISCUSSION TITLE
non-native positional aesthetics - beyond x and y

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#40] [General]
I think it's good to collect some non x and y positional aesthetics examples

Just to start: 

aes(x0 = x0, y0 = y0, r = r) in Extending your ability https://www.youtube.com/watch?v=uj7A3i2fi54

Where a non x and y aes determines x and y; dependency on other observations for placement.

aes(area = volume)                         in treemapify
aes(id = category, area = volume) in ggcirclepack
aes(label = word, size = speakers) in ggwordcloud

Where a non x and y aes determines x and y in pretty deterministic way (no dependency between observations)

aes(date = date) in ggcalendar.
aes(state = my_states_var) in ggusa (a hypothetical package)

**Comment by yjunechoe** [comment]
How do you define "positional aesthetics" in this case? Because I can think of at least four (what I think of when I hear that term is probably close to 3):

1) Very loosely, mappings in `aes()` that can affect the **final positioning of graphical elements** (ex: things that have consequences at the drawing stage like via `makeContent()`, such as the `spacing` aesthetic of `geom_textpath()` - i think??)
2) Somewhat loosely, mappings in `aes()` that can affect the **positioning of geoms on the data scale** such that the consequences of positioning are representable in the layer data  (like treemapify's `area`, which is presumably handled in Stat compute)
3) More strictly, mappings in `aes()` which are **simple reparameterizations of `x` and `y`** that can be resolved at the level of `$setup_data()` (e.g., `x0` and `xend`) vs. (e.g., `area`, `volume`, `date`)
4) Very strictly, arguments or mappings in `aes()` that gets **passed into the `Position` ggproto** to transform (including "identity") values in the `x` and `y` columns of the layer data (so like a bunch of Positions in {ggpp})

I should clarify that as I understand these definitions, these thresholds hierarchical, so if you include (3) you probably include (3,4), etc.

**Reply by yjunechoe** [reply]
Also I actually see `x0` and `y0` listed in `ggplot2:::ggplot_global`, so those are arguably "native" to ggplot 😁 

```r
ggplot2:::ggplot_global$x_aes
#>  [1] "x"          "xmin"       "xmax"       "xend"       "xintercept"
#>  [6] "xmin_final" "xmax_final" "xlower"     "xmiddle"    "xupper"    
#> [11] "x0"

ggplot2:::ggplot_global$y_aes
#>  [1] "y"          "ymin"       "ymax"       "yend"       "yintercept"
#>  [6] "ymin_final" "ymax_final" "lower"      "middle"     "upper"     
#> [11] "y0"
```

**Reply by EvaMaeRey** [reply]
I'm not even sure! But Probably 2... If I wrote 'non-native positional aesthetic mappings' in the title, then maybe I'd narrow your thinking to 2?

**Reply by EvaMaeRey** [reply]
Interesting about x0 and y0!  I guess it's not strictly a non-native case.  But the usage feels non-native we can imagine p0 and q0 being the input aes in geom_circle, right? not intuitive but get same result.  I guess x0 and y0 are used with some Geom that I'm unfamiliar with.

**Reply by yjunechoe** [reply]
In spirit of usage they're the "same", but I think the technical difference is that if you named them `p0` and `q0`, that's not just a name change - ggplot will no longer recognize them as aesthetics that are targets of, e.g., x/y-scale transformations. So the fact that the literal names `x0` and `y0` are natively registered/recognized in ggplot matters a lot - I don't think Thomas could've fit his demo in the talk if he made that seemingly trivial choice of renaming `x0` and `y0` to `p0` and `q0`!

**Reply by teunbrand** [reply]
I guess `theta` and `r` are non-native positional aesthetics, that only have meaning in context of `coord_polar()`/`coord_radial()`

**Reply by EvaMaeRey** [reply]
hmmm.... So `x0` and `y0` v `p0` and `q0`, would matter for rendering the StatCircle + GeomPolygon layer would be observable with scale transformations?

**Comment by teunbrand** [comment]
A few weeks ago I had some fun un-hardcoding the `x` and `y` aesthetics: https://github.com/tidyverse/ggplot2/pull/6080.
However, I don't think this'll get merged as it affects to many method signatures.

**Reply by yjunechoe** [reply]
Oh BTW completely unrelated fun fact - I see:

```r
z <- outer(x, y, f)
...
as.vector(z)
```

Did you know that `as.vector(outer(...))` is `kronecker(...)`? I just learned of `kronecker()` like 2 days ago and what a weird find! There's also even an infix version `%x%` which does the default `FUN = "*"` - like what the heck!?

**Reply by teunbrand** [reply]
That function never made any sense to me, at least when using `outer()` I know what I'm doing.

As an aside; I hate when stuff is named after people in sciency stuff. 
Just give it some semi-informative name instead of some thing that I have to remember by storing it in rote memory.
Like, call it a 'rank inequality test' instead of a 'Kolmogorov-Smirnov test'.

I think there should be a `t()` somewhere in that reconstruction.
``` r
x <- 1:3
y <- 3:1

as.vector(outer(x, y))
#> [1] 3 6 9 2 4 6 1 2 3
kronecker(x, y)
#> [1] 3 2 1 6 4 2 9 6 3
as.vector(t(outer(x, y)))
#> [1] 3 2 1 6 4 2 9 6 3
```

<sup>Created on 2024-09-26 with [reprex v2.1.1](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
We also call it 'the grammar of graphics', not 'Wilkinson syntax' 👍 
(I apologise for not recognising the linguistical subtleties between grammar/syntax, June!)

**Reply by yjunechoe** [reply]
Oh good point about `t()` (or I guess `kronecker(y,x)` - probably how I did it before I forgot about this fact 😅)

> As an aside; I hate when stuff is named after people in sciency stuff.
> Just give it some semi-informative name instead of some thing that I have to remember by storing it in rote memory.

Strong agree!!! I feel like they could've also just merged the `kronecker()` behavior into `outer()` and give it some extra arguments.

Good candidate for a Tidy Design Principles case study? 👀 https://design.tidyverse.org/

**Reply by yjunechoe** [reply]
> (I apologise for not recognising the linguistical subtleties between grammar/syntax, June!)

Don't worry - I distance myself more from the syntacticians (as a non-syntactician linguist) than a non-linguist ever possibly can 😆😈

**Comment by EvaMaeRey** [comment]
I think I've used the phrase 'hard coded' in Stat definition when for creating an x y column via compute_* vs. defaulting to x and y w/ after_stat(x_pos), but I don't think this is the hard coding discussed in https://github.com/tidyverse/ggplot2/pull/6080, which is coord-level?

**Reply by teunbrand** [reply]
It means different things in different contexts. The PR mucks with coords layouts and facets.

**Comment by EvaMaeRey** [comment]
I was a little surprise that I didn't manage to set default aes for x and y (errors unless compute_group output data frame contains x and y as output) but can for label.  It seems like x and y are somehow *more* required than label?  I thought I might be able to only default-set x and y, and let the user change if they prefer.

https://github.com/EvaMaeRey/easy-geom-recipes/discussions/11#discussioncomment-10800284 @yjunechoe the overall idea is to swap mean visualization for some linear modeling example - and would be used to introduce default_aes+after_stat!  :-)

``` r
library(tidyverse)
compute_group_xmean <- function(data, scales){

data |>
summarize(mean_x = mean(x),
xintercept = mean_x,
y = I(.5),
x = mean_x)

}

StatXMean <- ggproto("StatXMean", Stat,
compute_group = compute_group_xmean,
default_aes = aes(label = round(after_stat(mean_x), 2)),
required_aes = "x")

cars|>
ggplot() +
aes(speed, dist) +
geom_point() +
layer(geom = GeomVline, stat = StatXMean, position = "identity") +
geom_label(stat = StatXMean)
```

![](https://i.imgur.com/1EDKltG.png)

``` r

# Kind of surprised that following doesn't:

library(tidyverse)
compute_group_xmean <- function(data, scales){

data |>
summarize(mean_x = mean(x),
          xintercept = mean_x, 
          mean_range_y = mean(range(y)))

}

StatXMean <- ggproto("StatXMean", Stat,
                     compute_group = compute_group_xmean,
                     default_aes = aes(label = round(after_stat(mean_x), 2),
                                       y = I(.5),
                                       x = after_stat(mean_x)))

cars|>
  ggplot() + 
  aes(speed, dist) +
  geom_point() + 
  layer(geom = GeomVline, stat = StatXMean, position = "identity") +
  geom_label(stat = StatXMean)
#> Warning: The following aesthetics were dropped during statistical transformation: x and
#> y.
#> ℹ This can happen when ggplot fails to infer the correct grouping structure in
#>   the data.
#> ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
#>   variable into a factor?
#> The following aesthetics were dropped during statistical transformation: x and
#> y.
#> ℹ This can happen when ggplot fails to infer the correct grouping structure in
#>   the data.
#> ℹ Did you forget to specify a `group` aesthetic or to convert a numerical
#>   variable into a factor?
#> Error in `geom_label()`:
#> ! Problem while setting up geom.
#> ℹ Error occurred in the 3rd layer.
#> Caused by error in `compute_geom_1()`:
#> ! `geom_label()` requires the following missing aesthetics: x and y.
```

<sup>Created on 2024-09-30 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

**Reply by teunbrand** [reply]
The `default_aes` field in Stat classes is a bit finnicky: https://github.com/tidyverse/ggplot2/issues/3860

**Reply by yjunechoe** [reply]
I went down a rabbit hole and found your comment on a related PR:

> Also came across this:
> 
> https://github.com/tidyverse/ggplot2/blob/d31c051bb87cd5d5f02e8d19b5c002ad2a129d6c/R/stat-density-2d.R#L124
> 
> Where it seems weird to me that the stat declares aesthetics for the geom.

I got curious whether I can scan extension packages on CRAN for unusual default_aes mappings. I'm not sure the regex is most optimal/has best coverage, but I got something close: https://github.com/search?q=language%3Ar+user%3Acran+%2Fggproto%5C%28.*%3F%2C%5Cs*Stat%5Cw*%2C.*%3F%28%5Cn%7C.%29*default_aes%28%5Cn%7C.%29*%3F%5C%29%2F&type=code

```
language:r user:cran /ggproto\(.*?,\s*Stat\w*,.*?(\n|.)*default_aes(\n|.)*?\)/
```

That returns ~100 hits, and I see some interesting `default_aes` in a Stat like:

```r
# ggtern::StatTriTern/StatHexTern
aes(value = 1)

# ggalt::StatAsh
aes(colour = NA, fill = "gray20", size = 0.5, linetype = 1, alpha = NA)

# ggalt::StatBkde2d
aes(colour = "#3366FF", size = 0.5)

# ggpubr::StatCompareMultipleMeans
aes(wid = NULL)
```

I also see in ggplot2 something unusual in StatDensity - I don't fully understand the fill/weight defaults but probably just lack of my understanding

```r
ggplot2::StatDensity$default_aes
#> Aesthetic mapping: 
#> * `x`      -> `after_stat(density)`
#> * `y`      -> `after_stat(density)`
#> * `fill`   -> NA
#> * `weight` -> NULL
```

**Reply by teunbrand** [reply]
My guess these are just vestigial for some reason. If there is no harm in these useless things lingering, it can be many years before somebody notices that something is weird.

For example the `weight = NULL` was likely a way to delete/clean up a `weight` column after computation. This naturally happens for any non-constant weight column, but with the constant-preserving behaviour of stat computation, a simple `weight = 1L` might result in a misleading column in the data sent to the geom.

**Reply by yjunechoe** [reply]
> For example the weight = NULL was likely a way to delete/clean up a weight column after computation. This naturally happens for any non-constant weight column, but with the constant-preserving behaviour of stat computation, a simple weight = 1L might result in a misleading column in the data sent to the geom.

Ah! And I suppose sometimes the weight column persisting after compute is sometimes useful, so you wouldn't want to hard code dropping of weight in the compute (and so have it be "toggled off" via the `aes()` instead)?

I dug in a bit more to where `StatDensity` is used and perhaps the design is due to the fact that `StatDensity` is recycled across `geom_density()` with `GeomDensity` default and `stat_density()` with `GeomArea` default, where only `GeomDensity` understands the `weight` aesthetic. So I suppose having it suppress, by default, the `weight` variable that it computes, makes it play nice with more geoms that it's paired with?
